### PR TITLE
feat(awscdk): add plugins to CDK typescript app (#1244)

### PR DESCRIPTION
**Pull request by tiararodney**
This enables CDK plugins to be provided via the rendered `cdk.json`.

```javascript
new awscdk.AwsCdkTypeScriptApp({
  name: 'test',
  defaultReleaseBranch: 'main',
  cdkVersion: '1.133.0',
  cdkPlugins: ['cdk-assume-role-credential-plugin'],
});
```

plugins will be added as a devDependency to the package manifest and support all
common projen specifiers:

- plain ('cdk-assume-role-credential-plugin')
- versioned (foo@^1.0.0)
- scoped (@foo/foo)
- versioned and scoped (@foo/bar@^1.0.0)
- local (foo@/my/local/path/index.js)

resolves: #1244

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/awscdk/awscdk-app-ts.ts [hunk 1]
```diff
@@ -98,6 +98,23 @@ export class AwsCdkTypeScriptApp extends TypeScriptAppProject {
       throw new Error('Expecting tsconfig.json');
     }
 
+    // add CDK plugins as dependencies
+    const cdkPlugins = ( options.cdkPlugins ?? [] ).map( ( plugin:string ) => {
```
Comment (eladb) [inline src/awscdk/awscdk-app-ts.ts:102]: Please format according to code style

File: src/awscdk/awscdk-deps.ts [hunk 2]
```diff
@@ -77,6 +77,13 @@ export interface AwsCdkDepsCommonOptions {
    * @deprecated For CDK 2.x use 'devDeps' instead
    */
   readonly cdkTestDependencies?: string[];
+
+  /**
+   * AWS CDK plugins that extend the CDK features.
+   *
+   * Plugins are regular node packages
```
Comment (eladb) [inline src/awscdk/awscdk-deps.ts:84]: Add `@default`

File: src/awscdk/awscdk-app-ts.ts (hunk 1 above)
Comment (tiararodney) [reply src/awscdk/awscdk-app-ts.ts]: https://github.com/tiararodney/projen/commit/bdcb0b09e71b5a417ed19b3efb1d3a48e949d8e8

File: src/awscdk/awscdk-deps.ts (hunk 2 above)
Comment (tiararodney) [reply src/awscdk/awscdk-deps.ts:84]: https://github.com/tiararodney/projen/commit/bdcb0b09e71b5a417ed19b3efb1d3a48e949d8e8

File: src/awscdk/awscdk-app-ts.ts [hunk 3]
```diff
@@ -98,6 +98,23 @@ export class AwsCdkTypeScriptApp extends TypeScriptAppProject {
       throw new Error('Expecting tsconfig.json');
     }
 
+    // add CDK plugins as dependencies
+    const cdkPlugins = (options.cdkPlugins ?? []).map((plugin: string) => {
+
+      this.addDevDeps( plugin );
+
+      //field prefix (scope and semver) count
```
Comment (eladb) [inline src/awscdk/awscdk-app-ts.ts:106]:
```suggestion
      // field prefix (scope and semver) count
```

File: src/awscdk/awscdk-app-ts.ts [hunk 4]
```diff
@@ -98,6 +98,23 @@ export class AwsCdkTypeScriptApp extends TypeScriptAppProject {
...
+      const fpc = (plugin.match(/@/g) || []).length;
```
Comment (eladb) [inline src/awscdk/awscdk-app-ts.ts:107]:
```suggestion
      const fpc = (plugin.match(/@/g) ?? []).length;
```

File: src/awscdk/awscdk-app-ts.ts [hunk 5]
```diff
@@ -98,6 +98,23 @@ export class AwsCdkTypeScriptApp extends TypeScriptAppProject {
...
+
+      if (fpc > 1 || (fpc === 1 && plugin[0] != '@')) {
+
+        //remove the semver prefix
```
Comment (eladb) [inline src/awscdk/awscdk-app-ts.ts:111]:
```suggestion
        // remove the semver prefix
```

File: src/awscdk/awscdk-deps.ts [hunk 6]
```diff
@@ -77,6 +77,15 @@ export interface AwsCdkDepsCommonOptions {
    * @deprecated For CDK 2.x use 'devDeps' instead
    */
   readonly cdkTestDependencies?: string[];
+
+  /**
+   * AWS CDK plugins that extend the CDK features.
```
Comment (eladb) [inline src/awscdk/awscdk-deps.ts:82]:
```suggestion
   * AWS CDK CLI plugins that extend the CDK features.
```

File: src/awscdk/awscdk-deps.ts [hunk 7]
```diff
@@ -77,6 +77,15 @@ export interface AwsCdkDepsCommonOptions {
...
+   *
+   * Plugins are regular node packages
+   *
```
Comment (eladb) [inline src/awscdk/awscdk-deps.ts:85]: Can you add a "@link" to the relevant AWS CDK docs here?

File: src/awscdk/cdk-config.ts [hunk 8]
```diff
@@ -58,6 +58,12 @@ export interface CdkConfigCommonOptions {
    * @default []
    */
   readonly watchExcludes?: string[];
+  /**
```
Comment (eladb) [inline src/awscdk/cdk-config.ts:61]:
```suggestion

  /**
```

File: test/awscdk/awscdk-app.test.ts [hunk 9]
```diff
@@ -170,3 +170,115 @@ test('CDK v1 usage', () => {
...
+    const project = new awscdk.AwsCdkTypeScriptApp({
+      name: 'test',
+      defaultReleaseBranch: 'main',
+      cdkVersion: '1.133.0',
+      cdkPlugins: plugins,
+    });
+
+    const files = synthSnapshot(project);
+
+    expect( files['cdk.json'].plugin ).toStrictEqual(['foo']);
```
Comment (eladb) [inline test/awscdk/awscdk-app.test.ts:207]: Plz fix formatting

**Review by eladb — CHANGES_REQUESTED** [review]

**Comment by eladb** [comment]
## Please read this to avoid merge hell

Hola! apologies for hassle. I've enabled Prettify as part of #1453. This means that most of the source files have changed and you will likely see quite a lot of merge conflicts when you try to merge from `main`.

*But there's an easy way to handle this.*

First, merge your branch from this tag:

```sh
$ git fetch origin
$ git merge 1020d0ad3f95bb4e1201f8640d440ffb43ccf230
```

Edit the `.projenrc.js` file in your branch and add this to **line 57** (right
after `codeCov: true`):

```js
  prettier: true,
```

Now, run:

```sh
$ git clean -fdx .
$ yarn
$ bash projen.bash
$ yarn eslint
$ git add .
$ git commit -m 'format code'
```

This will reformat all the code in your branch to match `main`.

Now, merge from main:

```sh
$ git merge origin/main
```

Sorry again for the inconvenience...

File: src/awscdk/awscdk-deps.ts (hunk 7 above)
Comment (pflorek) [reply src/awscdk/awscdk-deps.ts:85]: I assume there is no documentation except https://github.com/aws/aws-cdk/blob/master/packages/aws-cdk/lib/plugin.ts and there is an argument on the command line

**Pull request closed without merging**